sql/log.cc

```diff
@@ -12914,3 +13024,221 @@ void wsrep_register_binlog_handler(THD *thd, bool trx)
+  */
+  if (unlikely(cache_data->cache_log.disk_writes == 0))
+    return false;
+
+  return true;
+}
+
+bool Binlog_cache_free_flush::commit(MYSQL_BIN_LOG::group_commit_entry *entry)
+{
+  bool check_purge= false;
+  THD *thd= entry->thd;
+  binlog_cache_mngr *cache_mngr= entry->cache_mngr;
+  binlog_cache_data *cache_data = cache_mngr->get_binlog_cache_data(true);
+  if (unlikely(!entry->using_trx_cache || cache_data->empty()))
+    cache_data= cache_mngr->get_binlog_cache_data(false);
+
+  /* Sync the temp file before enter log_lock to avoid holding the lock long */
+  if (cache_data->sync_tmp_file())
+    return false;
+
+  thd->wait_for_prior_commit();
+
+  // It will be release by trx_group_commit_leader
+  mysql_mutex_lock(&mysql_bin_log.LOCK_log);
+
+  enum enum_binlog_checksum_alg expected_alg=
+      mysql_bin_log.checksum_alg_reset != BINLOG_CHECKSUM_ALG_UNDEF
+          ? mysql_bin_log.checksum_alg_reset
+          : (enum_binlog_checksum_alg) binlog_checksum_options;
+
+  /*
+    In legacy mode, all events should has a valid position this done by
+    updating log_pos field when writing events from binlog cache to binlog
+    file. Thus rename binlog cache to binlog file is not supported in legacy
+    mode.
+
+    if the cache's checksum alg is not same to the binlog's checksum, it needs
+    to recalculate the checksum. Thus rename binlog cache to binlog file is
+    not supported.
+  */
+  if (!mysql_bin_log.is_open() || opt_binlog_legacy_event_pos ||
+      (expected_alg != cache_data->checksum_opt))
+  {
+    mysql_mutex_unlock(&mysql_bin_log.LOCK_log);
+    return false;
+  }
+
+  m_entry= entry;
+  m_replaced= false;
+  m_cache_data= cache_data;
+  ulong prev_binlog_id= mysql_bin_log. current_binlog_id;
+  /*
+    It will call replace_binlog_file() to rename the transaction's binlog cache
+    to the new binlog file.
+  */
+  if (mysql_bin_log.rotate(true, &check_purge, true /* is_free_flush */))
+  {
+    DBUG_ASSERT(!m_replaced);
+    DBUG_ASSERT(!mysql_bin_log.is_open());
+  }
+
+  DBUG_EXECUTE_IF("binlog_cache_free_flush_crash_after_rotate",
+                  DBUG_SUICIDE(););
+
+  if (!m_replaced)
+  {
+    mysql_mutex_unlock(&mysql_bin_log.LOCK_log);
+    if (check_purge)
+      mysql_bin_log.checkpoint_and_purge(prev_binlog_id);
+    return false;
+  }
+
+  mysql_bin_log.trx_group_commit_leader(entry, true /* is_free_flush */);
+
+  if (check_purge)
+    mysql_bin_log.checkpoint_and_purge(prev_binlog_id);
+  return true;
+}
+
+bool Binlog_cache_free_flush::replace_binlog_file()
+{
+  size_t binlog_size= my_b_tell(&mysql_bin_log.log_file);
+  size_t required_size= binlog_size;
+  // space for Gtid_log_event
+  required_size+= LOG_EVENT_HEADER_LEN + Gtid_log_event::max_data_length +
+                  BINLOG_CHECKSUM_LEN;
+
+  if (DBUG_IF("simulate_required_size_too_big") ||
+      required_size > m_cache_data->file_reserved_bytes())
+  {
+    sql_print_information("Could not rename binlog cache to binlog, "
```

Comment:
This case should be unlikely, so 1) I'd wrap the `if` condition check with `unlikely()` and 2) can we just revert to the old behavior in this case? We shouldn't have yet written the `rotate` event, so we should be in a consistent state.

Reply:
If error happens here, binlog system will be closed. That is not what I expected. In fact, replace_binlog_file() is called after written rotate event. the old file has been closed and  the new file has been created. In the situation, it will go finish the rotate. But the transaction will go to the original path to copy its cache to the new binlog file. The only different is that there is an extra rotate.

Reply:
I didn't use unlikely in this function, that is because this function is called not often.